eyeshade/workers/reports.js

```diff
@@ -815,6 +816,153 @@ const date2objectId = (iso8601, ceilP) => {
                        (ceilP ? 'ffffffffffffffff' : '0000000000000000'))
 }
 
+/**
+ * A referral statement consists of entries describing accrued earnings from referrals,
+ * past successful settlements (payouts), and a balance entry showing a publishers current
+ * settlement balance from referrals.
+ **/
+const referralStatement = async (debug, runtime, owner, summaryP) => {
+  const referrals = runtime.database.get('referrals', debug)
+  const settlements = runtime.database.get('settlements', debug)
+
+  if (!summaryP) {
+    throw new Error('non summary not currently supported')
+  }
+
+  const statements = []
+  const statementTemplate = {
+    referrals: { summary: {}, entries: {} },
+    settlements: { summaries: [], entries: {} },
+    balance: {}
+  }
+
+  const referralFilter = { probi: { $gt: 0 }, altcurrency: { $eq: altcurrency }, exclude: false }
+  const settlementFilter = { probi: { $gt: 0 }, altcurrency: { $eq: altcurrency }, type: { $eq: 'referral' } }
+  if (owner) {
+    referralFilter.owner = { $eq: owner }
+    settlementFilter.owner = { $eq: owner }
+  }
+
+  const referralTotals = await referrals.aggregate([
+    { $match: referralFilter },
+    {
+      $group: {
+        _id: '$publisher',
+        count: { $sum: 1 },
+        probi: { $sum: '$probi' }
+      }
+    }
+  ])
+  referralTotals.forEach((total) => {
+    total.publisher = total._id
+    total.probi = new BigNumber(total.probi)
```

Comment:
it is probably safer to use `...toString()` when taking a `Decimal128` to create a `BigNumber` -- if the argument to `BigNumber` is already a string, then adding a `.toString()` has, at most, a negligible performance impact. in this case, i think the argument isn't going to be a string.